src/utils/allowHook.ts

```diff
@@ -0,0 +1,101 @@
+import * as fs from 'fs';
+import * as path from 'path';
+
+import * as execa from 'execa';
+import * as express from 'express';
+
+import {CLIArguments} from '../types';
+
+import {getRootDirectory} from './getRootDirectory';
+
+type HookableFunction = (...args: any[]) => any;
+const hooks: {[key: string]: HookableFunction} = {};
+
+export enum Hooks {
+    /**
+     * Register middleware with the 'npm start' server.
+     *
+     * Any middleware added via this hook will take precedence over the "built-in" middleware.
+     */
+    APP_MIDDLEWARE,
+    /**
+     * Register middleware with the local server that runs during integration tests.
+     *
+     * Any middleware added via this hook will take precedence over the "built-in" middleware.
+     */
+    TEST_MIDDLEWARE,
+    /**
+     * Hook to override default values for the `npm start` CLI options.
+     *
+     * Note that using this hook may make some of the text in `npm start --help` inaccurate.
+     */
+    DEFAULT_ARGS,
+}
+
+export interface HooksAPI {
+    [Hooks.DEFAULT_ARGS]: () => Partial<CLIArguments>;
+    [Hooks.APP_MIDDLEWARE]: (app: express.Express) => void;
+    [Hooks.TEST_MIDDLEWARE]: (app: express.Express) => void;
+}
+
+export function loadHooks(): void {
+    const hooksPathSrc = path.resolve(getRootDirectory(), 'hooks.ts');
+    const hooksPathOut = path.resolve(getRootDirectory(), 'hooks.js');
+    if (fs.existsSync(hooksPathSrc)) {
+        // Build hooks if first-run, or source was modified
```

Comment:
Could this be done more easily using the typescript API directly, instead of invoking it with the cli?

Reply:
Yeah, but from past experience getting sane error output is fiddly. I went with the one-liner approach...

See the "minimal" code sample: https://github.com/microsoft/TypeScript/wiki/Using-the-Compiler-API#a-minimal-compiler

Reply:
Hmm, ok. I'd still rather do it that way, but probably not worth the effort right now.